dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java

```diff
@@ -0,0 +1,673 @@
+        len = dis.readInt();
+        List<ConfigID> ids = new ArrayList<>();
+        for (int i = 0, j = 0, cursor = 0; i < len + cursor; i++, j++) {
+            if (cursor == 0) {
+                cursor = dis.readInt();
+                j = 0;
+            }
+            String id = readPossiblyNullString(dis);
+            String configPid = dis.readUTF();
+            String child = readPossiblyNullString(dis);
+            int parentId = dis.readInt();
+            ConfigID parent = null;
+            if (parentId > -1) {
+                parent = ids.get(parentId);
+            }
+            ConfigID thisId = new ConfigID(parent, configPid, id, child);
+            ids.add(thisId);
+
+            if (j == cursor) {
+                len = len + cursor;
+                cursor = 0;
+                references.add(thisId);
+            }
+        }
+        readMap(dis, toMapOrDictionary(dict));
+        return location;
+
+    }
+
+    private static String readPossiblyNullString(DataInputStream dis) throws IOException {
+        if (dis.readBoolean()) {
+            return dis.readUTF();
+        }
+        return null;
+    }
+
+    public static void store(File configFile, Collection<? extends ExtendedConfiguration> configs) throws IOException {
+        try (DataOutputStream dos = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(configFile, false)))) {
+            dos.writeByte(VERSION);
+            dos.writeInt(configs.size());
+            for (ExtendedConfiguration config : configs) {
+                Dictionary<String, ?> properties = config.getReadOnlyProperties();
+                Set<ConfigID> references = config.getReferences();
+                Set<String> uniqueVars = config.getUniqueVariables();
+                String bundleLocation = config.getBundleLocation();
+                store(dos, properties, bundleLocation, references, uniqueVars);
+            }
+        }
+    }
+
+    public static void store(DataOutputStream dos, Dictionary<String, ?> properties, String bundleLocation,
+                             Set<ConfigID> references, Set<String> uniqueVars) throws IOException {
+
+        writePossiblyNullString(dos, bundleLocation);
+        if (uniqueVars != null) {
+            dos.writeInt(uniqueVars.size());
+            for (String var : uniqueVars) {
+                dos.writeUTF(var);
+            }
+        } else {
+            dos.writeInt(0);
+        }
+        if (references != null) {
+            dos.writeInt(references.size());
+            int count = 0;
+            Map<ConfigID, Integer> writtenConfigIds = new HashMap<>();
+            for (ConfigID id : references) {
+                count = writeConfigID(dos, id, count, 0, writtenConfigIds);
+            }
+        } else {
+            dos.writeInt(0);
+        }
+        writeMap(dos, toMapOrDictionary(properties));
+    }
+
+    public static void readMap(DataInputStream dis, MapIterable props) throws IOException {
```

Comment:
does this need to be public?

Reply:
Yes for usage by the variable registry cache which needs to persist a map with the same types as config.